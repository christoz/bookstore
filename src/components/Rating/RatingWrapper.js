import styled from 'styled-components';

const RatingWrapper = styled.div`
	display: flex;
	align-items: center;

	[role="img"] {
		cursor: pointer;
		font-size: 1.6rem;
	}
`;

export default RatingWrapper;
