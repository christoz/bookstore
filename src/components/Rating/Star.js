import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledStar = styled.span`
	display: inline-block;
	color: ${props => props.selected ? '#fc6' : '#ccc'};
`;

const Star = ({ onSelect, selected }) => {
	return (
		<StyledStar
			role="img"
			aria-label="star"
			selected={selected}
			onClick={onSelect}
		>
			â˜…
		</StyledStar>
	)
};

Star.propTypes = {
	onSelect: PropTypes.func,
	selected: PropTypes.bool
}

export default Star;
