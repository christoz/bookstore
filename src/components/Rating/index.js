import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';

import Star from './Star';
import Wrapper from './RatingWrapper';

const Rating = ({ onRate, rating, stars }) => {
	const [rateValue, setRateValue] = useState(rating);

	const onSetRate = useCallback(function (value) {
		setRateValue(value)
		// Optional call to lift up rate to parent component
		if (onRate) {
			onRate(value)
		}
	}, [onRate]);

	return (
		<Wrapper>
			{
				[...Array(stars)].map((_, idx) => {
					return (
						<Star
							key={idx}
							selected={idx < rateValue}
							onSelect={() => onSetRate(idx + 1)}
						/>
					)
				})}
		</Wrapper>
	);
};

Rating.propTypes = {
	stars: PropTypes.number,
	rating: PropTypes.oneOf([0, 1, 2, 3, 4, 5]),
	onRate: PropTypes.func
};

export default Rating;
