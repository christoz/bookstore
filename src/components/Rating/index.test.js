import React from 'react';
import ReactDom from 'react-dom';
import Rating from './index';

describe('<Rating /> spec', () => {
	it('renders without problems', () => {
		const div = document.createElement('div');
		ReactDom.render(<Rating />, div);
	});

	it('should render 5 stars', () => {
		const div = document.createElement('div');
		ReactDom.render(<Rating stars={5} />, div);

		const numOfStars = div.querySelectorAll('[role="img"]').length;
		expect(numOfStars).toBe(5);
	});
});
