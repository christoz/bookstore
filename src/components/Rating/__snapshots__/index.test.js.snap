// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rating /> spec renders without problems 1`] = `
<div>
  <div
    class="sc-AxirZ jKLVTp"
  >
    <span
      aria-label="star"
      class="sc-AxjAm fiNLJi"
      role="img"
    >
      ★
    </span>
  </div>
</div>
`;

exports[`<Rating /> spec should render 5 stars 1`] = `
<div>
  <div
    class="sc-AxirZ jKLVTp"
  >
    <span
      aria-label="star"
      class="sc-AxjAm fiNLJi"
      role="img"
    >
      ★
    </span>
  </div>
</div>
`;
